File: api/app/javascript/lib/typography.js
// The part of the SmartyPants typography that changes the LENGTH of a post.
//
// ⚠️ **This file is for the COUNT, and it must never render text that a person reads.** It writes
// the ellipsis and the two dashes and it leaves every quotation mark straight. The server, through
// `Typography` and `MarkdownHelper#smartypants`, is the one place that writes the true text.
//
// ⚠️ **A quotation mark is why that split is safe.** `"` becomes `“` or `”` and `'` becomes `’`,
// and each of those is ONE character in place of one. Thus the direction of a quotation mark, which
// is the hard half of SmartyPants and the half that no small copy can match, cannot change a count.
// A measurement against Redcarpet gave the same length for every quotation shape.
//
// The rules below are what is left, and each one is context-free:
//
//   ...  or  . . .   ->  …     (three characters, or three with ONE space between each, become one)
//   ---                ->  —
//   --                 ->  –
//
// ⚠️ The dashes run LONGEST FIRST, as Redcarpet does. Four dashes give `—-` and five give `—–`.
//
// ⚠️ An address goes through with NO change, exactly as `Typography` leaves it alone: a link that
// holds `--` must keep it, and a count that shortened one would be smaller than the server's.

import { TOKEN_SOURCE, URL_SOURCE } from "./social_mentions";

const ELLIPSIS = /\.\.\.|\. \. \./g;
const EM_DASH = /---/g;
const EN_DASH = /--/g;
// The marks and the fractions that SmartyPants writes as one character.
const COPYRIGHT = /\(c\)/gi;
const REGISTERED = /\(r\)/gi;
const TRADEMARK = /\(tm\)/gi;
const FRACTIONS = /(?<![\d/])(1\/4|1\/2|3\/4)(?![\d/])/g;
const FRACTION_CHARS = { "1/4": "¼", "1/2": "½", "3/4": "¾" };

/**
 * Applies the rules that change the length of a post.
 *
 * ⚠️ For the count only. Refer to the ⚠️ at the top of this file.
 * @param {string} text
 * @returns {string} The text at the length that the server will post.
 */
export function applyLengthRules(text) {
  const body = text ?? "";
  if (!body) return body;

  let out = "";
  let last = 0;

  for (const [start, end] of maskedRanges(body)) {
    out += convert(body.slice(last, start)) + body.slice(start, end);
    last = end;
  }

  return out + convert(body.slice(last));
}

/**
 * @param {string} chunk A piece of the text that holds no address.
 * @returns {string}
 */
function convert(chunk) {
  return chunk
    .replace(ELLIPSIS, "…")
    .replace(EM_DASH, "—")
    .replace(EN_DASH, "–")
    .replace(COPYRIGHT, "©")
    .replace(REGISTERED, "®")
    .replace(TRADEMARK, "™")
    .replace(FRACTIONS, (match) => FRACTION_CHARS[match]);
}

/**
 * Each address and each mention token, in order. ⚠️ A handle can hold `--`, and the server keeps
 * it: `Typography#masked_ranges` masks the same two shapes.
 * @param {string} text
 * @returns {Array<[number, number]>}
 */
function maskedRanges(text) {
  const urls = groupRanges(text, URL_SOURCE);
  const tokens = groupRanges(text, TOKEN_SOURCE).filter(
    ([start]) => !urls.some(([from, to]) => start >= from && start < to)
  );
  return [...urls, ...tokens].sort((a, b) => a[0] - b[0]);
}

/**
 * The character ranges of group 1 of a pattern.
 * @param {string} text
 * @param {string} source
 * @returns {Array<[number, number]>}
 */
function groupRanges(text, source) {
  const pattern = new RegExp(source, "g");
  const ranges = [];
  let match;

  while ((match = pattern.exec(text)) !== null) {
    const start = match.index + match[0].length - match[1].length;
    ranges.push([start, start + match[1].length]);
    if (match.index === pattern.lastIndex) pattern.lastIndex += 1;
  }
  return ranges;
}

File: api/app/services/typography.rb
# Applies the SmartyPants typography to the draft of a social post: curly quotation marks, a true
# ellipsis, and an en or em dash, from the characters that a person types on a keyboard.
#
# It is not an ApplicationService, because that base class is for HTTP integrations and this class
# makes no network call. Each method is a class method, thus a caller needs no instance.
#
# ⚠️ **It is the SAME typography as the blog**, through `MarkdownHelper#smartypants`. The owner
# writes for one voice, thus a post and the article that it links to must not use a different
# apostrophe. Do not write a second set of rules here.
#
# ⚠️ **A URL goes through with NO change**, and that is the whole reason that this class exists and
# the helper is not called directly. SmartyPants reads the characters of an address as punctuation:
# `example.com/a--b` becomes an en dash, `?q="a"` becomes curly quotation marks, and `/a...b`
# becomes an ellipsis. Each of those gives a link that is dead, and nothing reports it.
class Typography
  # ⚠️ It extends the helper, thus this class cannot hold a rule of its own. Refer to the ⚠️ above.
  extend MarkdownHelper

  # What each address becomes while SmartyPants reads the text. U+FFFC is OBJECT REPLACEMENT
  # CHARACTER, which is what it names: one character that stands for something that is not text.
  #
  # ⚠️ **The mask is necessary, and a split at each address is NOT enough.** SmartyPants decides the
  # direction of a quotation mark from the characters at each side of it, thus it must read the
  # whole sentence at one time. With a split, `He said "see <link> now"` opens the quotation and
  # never closes it: the two marks are in different pieces.
  PLACEHOLDER = "\uFFFC".freeze

  class << self
    # @param text [String, nil] The words that the owner wrote.
    # @return [String] The same words, with the typography of the site.
    def apply(text)
      # ⚠️ A draft of its own cannot hold the mask, or the addresses would go back in the wrong
      # places. That character is not text and it renders as nothing, thus this loses no words.
      text = text.to_s.delete(PLACEHOLDER)
      return text if text.empty?

      urls = []
      converted = convert(mask(text, urls))

      # ⚠️ **It gives the words back with NO typography when a mask went missing.** A post with a
      # straight apostrophe is a small thing, and an address that another address replaced is a
      # link that goes to the wrong page.
      return text unless converted.count(PLACEHOLDER) == urls.length

      index = -1
      converted.gsub(PLACEHOLDER) { urls[index += 1] }
    end

    private

    # Writes the text with one mask in place of each address, and collects those addresses.
    # @param text [String]
    # @param urls [Array<String>] Filled in, in the order that the addresses appear.
    # @return [String]
    def mask(text, urls)
      masked = +""
      last = 0

      url_ranges(text).each do |range|
        masked << text[last...range.begin] << PLACEHOLDER
        urls << text[range]
        last = range.end
      end

      masked << text[last..].to_s
    end

    # ⚠️ **SmartyPants writes HTML entities**, because it is a renderer of HTML: it gives `&rsquo;`
    # and not `’`. A post holds characters, thus the decode is necessary and not decoration.
    # `MarkdownHelper#markdown_to_plain_text` decodes for the same reason.
    # @param chunk [String]
    # @return [String]
    def convert(chunk)
      return chunk if chunk.empty?

      HTMLEntities.new.decode(smartypants(chunk).to_s)
    end

    # ⚠️ It uses the URL pattern of Bluesky, which is the one that decides a link facet. Thus a
    # string that becomes a link there is an address here, and one pattern holds that rule.
    # `SocialMentions` reads it for the same reason.
    #
    # ⚠️ It covers the address of a Markdown link as well. `](https://…)` and a `[name]: https://…`
    # definition each put the address after a character that is not a word character, thus the
    # pattern finds both.
    # @param text [String]
    # @return [Array<Range>] The character ranges of each address, in order.
    def url_ranges(text)
      ranges = []
      text.scan(Bluesky::URL_PATTERN) do
        start_char, end_char = Regexp.last_match.offset(1)
        ranges << (start_char...end_char)
      end
      ranges
    end
  end
end
